### PR TITLE
Add integration test to check that errors during deletion do not remove finalizer

### PR DESCRIPTION
**Pull request by MarcelMue**
towards https://github.com/giantswarm/giantswarm/issues/1595

### CODE REVIEW

File: controller/integration/test/controlflow/controlflow_test.go [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+	testWrapper := wrapper.Interface(nodeconfigWrapper)
+
+	testWrapper.MustSetup(testNamespace)
+	defer testWrapper.MustTeardown(testNamespace)
+
+	controller := testWrapper.Controller()
+
+	// We start the controller.
+	go controller.Boot()
```
Comment (MarcelMue) [inline controller/integration/test/controlflow/controlflow_test.go]: The real action starts here, the testcase is pretty straight forwards but a lot of checks in there to make sure everything is alright.

Steps taken here:
1. Boot framework
2. Create CR and check if the finalizer gets added
3. Set `ThrowError` to true
4. Delete CR and check that it's not removed and finalizer still exists
5. Set `ThrowError` to false
6. Check if the CR is completely gone

File: controller/integration/testresource/testresource.go [hunk 2]
```diff
@@ -32,33 +34,44 @@ func New(config Config) (*Resource, error) {
 		createCount: 0,
 		deleteCount: 0,
 		name:        config.Name,
+		throwError:  false,
 	}
 
 	return r, nil
 }
 
+func (r *Resource) CreateCount() int {
```
Comment (MarcelMue) [inline controller/integration/testresource/testresource.go]: Did some boyscouting and removed the `Get` prefix here.

File: controller/integration/testresource/testresource.go [hunk 3]
```diff
@@ -32,33 +34,44 @@ func New(config Config) (*Resource, error) {
 		createCount: 0,
 		deleteCount: 0,
 		name:        config.Name,
+		throwError:  false,
```
Comment (MarcelMue) [inline controller/integration/testresource/testresource.go]: Pretty simple concept: throw errors in all resource functions if set to `true`.

File: controller/integration/test/controlflow/controlflow_test.go [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+	}
+
+	// We verify, that our finalizer is still set.
+	if !reflect.DeepEqual(resultObjAccessor.GetFinalizers(), expectedFinalizers) {
+		t.Fatalf("finalizers == %v, want %v", resultObjAccessor.GetFinalizers(), expectedFinalizers)
+	}
+
+	// We set ThrowError to false, our finalizer should be removed with the next
+	// reconciliation now.
+	tr.ThrowError(false)
```
Comment (xh3b4sd) [inline controller/integration/test/controlflow/controlflow_test.go]: IMO we can drop the argument and only call `ThrowError` when we want the resource to throw an error. So the second time here is not necessary to call it at all.

File: controller/integration/test/controlflow/controlflow_test.go [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+	"github.com/giantswarm/operatorkit/controller/integration/testresource"
+	"github.com/giantswarm/operatorkit/controller/integration/wrapper"
+	"github.com/giantswarm/operatorkit/controller/integration/wrapper/nodeconfig"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+)
+
+// Test_Finalizer_Integration_Controlflow is a integration test to
+// check that errors during deletion prevent the finalizer from removal.
+func Test_Finalizer_Integration_Controlflow(t *testing.T) {
```
Comment (xh3b4sd) [inline controller/integration/test/controlflow/controlflow_test.go]: I like this test a lot. Understood it the first time reading. :D

File: controller/integration/test/controlflow/controlflow_test.go (hunk 4 above)
Comment (MarcelMue) [reply controller/integration/test/controlflow/controlflow_test.go]: I added the explicit argument for the toggle, so we can add some more complex scenarios when we have a working solution for the cancel context stuff.

I didn't want to add the current solution as it isn't working as you reported in https://github.com/giantswarm/giantswarm/issues/3018

File: controller/integration/test/controlflow/controlflow_test.go (hunk 1 above)
Comment (kopiczko) [reply controller/integration/test/controlflow/controlflow_test.go]: Pleas no error throwing in Go, returning only! :P

**Comment by MarcelMue** [comment]
I replaced `throw`with `return`, going with this. 
Ping me if anything else needs to be changed <3

**Pull request merged**